src/gpuarray_blas_cuda_cublas.c

```diff
@@ -528,6 +763,269 @@ static int dgemv(cb_order order, cb_transpose transA, size_t M, size_t N,
+                      size_t batchCount, int flags) {
+  cuda_context *ctx;
+  size_t t, i;
+  size_t ls[2], gs[2];
+  void *args[9];
+  gpudata *Aa, *xa, *ya;
+  int err;
+
+  if (flags != 0) return GA_INVALID_ERROR;
+  if (batchCount == 0) return GA_NO_ERROR;
+
+  if (alpha != 1.0 || beta != 1.0) return GA_UNSUPPORTED_ERROR;
+
+  if (M < 512) {
+    ls[0] = 32;
+    if (batchCount > 16)
+      ls[1] = 16;
+    else
+      ls[1] = batchCount;
+  } else {
+    ls[0] = 512;
+    ls[1] = 1;
+  }
+  gs[0] = (M + ls[0] - 1) / ls[0];
+  gs[1] = (batchCount + ls[1] - 1) / ls[1];
+  if (gs[0] * gs[1] / 65535) {
+    gs[1] = (65535 / gs[0]);
+  }
+
+  if (order == cb_c) {
+    t = N;
+    N = M;
+    M = t;
+    if (transA == cb_no_trans) {
+      transA = cb_trans;
+    } else {
+      transA = cb_no_trans;
+    }
+  }
+
+  ASSERT_BUF(A[0]);
+
+  ctx = A[0]->ctx;
+
+  cuda_enter(ctx);
+
+  {
+    float **T_l = alloca(sizeof(float *) * batchCount * 3);
+    const float **A_l = (const float **)T_l;
+    const float **x_l = (const float **)T_l + batchCount;
+    float **y_l = T_l + (batchCount * 2);
+
+    for (i = 0; i < batchCount; i++) {
+      ASSERT_BUF(A[i]);
+      ASSERT_BUF(x[i]);
+      ASSERT_BUF(y[i]);
+      cuda_wait(A[i], CUDA_WAIT_READ);
+      cuda_wait(x[i], CUDA_WAIT_READ);
+      cuda_wait(y[i], CUDA_WAIT_READ|CUDA_WAIT_WRITE);
+      A_l[i] = (float *)(A[i]->ptr + offA[i]);
+      x_l[i] = (float *)(x[i]->ptr + offX[i]);
+      y_l[i] = (float *)(y[i]->ptr + offY[i]);
+    }
+
+    Aa = cuda_ops.buffer_alloc(ctx, sizeof(float *) * batchCount, A_l,
+                               GA_BUFFER_INIT, &err);
+    if (Aa == NULL)
+      return err;
+    xa = cuda_ops.buffer_alloc(ctx, sizeof(float *) * batchCount, x_l,
+                               GA_BUFFER_INIT, &err);
+    if (xa == NULL) {
+      cuda_ops.buffer_release(Aa);
+      return err;
+    }
+    ya = cuda_ops.buffer_alloc(ctx, sizeof(float *) * batchCount, y_l,
+                               GA_BUFFER_INIT, &err);
+    if (ya == NULL) {
+      cuda_ops.buffer_release(Aa);
+      cuda_ops.buffer_release(xa);
+      return err;
+    }
+  }
+
+  args[0] = Aa;
+  args[1] = &lda;
+  args[2] = xa;
+  args[3] = &incX;
+  args[4] = ya;
+  args[5] = &incY;
+  args[6] = &batchCount;
+  args[7] = &M;
+  args[8] = &N;
+
+  if (transA == cb_no_trans) {
+    err = GpuKernel_call(&((blas_handle *)ctx->blas_handle)->sgemvBH_N_a1_b1_small, 2, ls, gs, 0, args);
+  } else {
+    err = GpuKernel_call(&((blas_handle *)ctx->blas_handle)->sgemvBH_T_a1_b1_small, 2, ls, gs, 0, args);
+  }
+
+  cuda_ops.buffer_release(Aa);
+  cuda_ops.buffer_release(xa);
+  cuda_ops.buffer_release(ya);
+
+  if (err != GA_NO_ERROR) {
+    cuda_exit(ctx);
+    return err;
+  }
+
+
+  for (i = 0; i < batchCount; i++) {
+    cuda_record(A[i], CUDA_WAIT_READ);
+    cuda_record(x[i], CUDA_WAIT_READ);
+    cuda_record(y[i], CUDA_WAIT_READ|CUDA_WAIT_WRITE);
```

Comment:
Out of curiosity, doing all those event, could this slow down computation? I won't delay this PR for that. Just wanted to know.